### PR TITLE
[Bench] Amortize the GDN ReplaySSM decode latency over the flush cycle

### LINKED ISSUE #38398: [Bug] bench_gdn_replayssm_decode.py never advances write_pos, so it times only the non-flush phase and overstates the speedup by up to 21%
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

## Describe the bug

`benchmark/kernels/attention/bench_gdn_replayssm_decode.py` creates
`write_pos = torch.zeros(B)` and never advances it, although the kernel wrapper's
own docstring (`fused_recurrent_linear_replayssm.py:472-473`) states that the
caller must:

> Allocates nothing persistent: the caller owns the ring tensors and is
> responsible for advancing / resetting `write_pos` (e.g. `(write_pos+1) % L`
> after each step)

`write_pos` appears exactly twice in the file, at line 97 where it is created and
line 115 where it is passed to the kernel. The timed closure `run_replay` (lines
101-118) does not touch it.

The kernel reads it twice:

```python
b_is_flush  = b_write_pos == MAX_CACHE_LEN - 1     # line 122
cache_valid = o_c < b_write_pos                    # line 129
```

With `write_pos` pinned at 0 and `L >= 2`, `b_is_flush` is false on every
iteration and `cache_valid` is the empty set. **The benchmark therefore never
writes the checkpoint and never replays the ring.** It measures one phase of the
L-phase cycle, the cheapest one, and reports it as the per-step cost.

### The two reported columns disagree with each other

This part needs no measurement. The traffic model at line 61 assumes a checkpoint
write once per L steps:

```python
replay = (state_elems * fp32) * (1 + 1.0 / L)   # read every step + write 1/L
```

The timing path performs that write **zero** times per L steps. In one line of
output, `state-traffic` is computed for 1-in-L writes while `speedup` is measured
for 0-in-L writes.

The labels claim the amortized quantity in both places. Line 149 prints
`per-step latency (ms); speedup = packed/replay`, and the module docstring says
"Compares per-step decode latency" and "The amortized state traffic ratio is
reported per L". Neither says the latency is steady-state only.

If timing only the non-flush phase is deliberate, then the column label and the
traffic model should say so, because as written the two contradict each other and
a reader has no way to tell which one to trust.

## Methodology

Phases are timed by pinning `write_pos` at each value in `[0, L)` and averaging,
rather than advancing the cursor inside the timed region, which would fold an
extra per-step kernel launch into the measurement.

The two agree. At `B=256, L=8`, a loop that really advances `write_pos` costs
**0.1935 ms/step**, against a phase-pinned mean of **0.1946 ms/step**, a **0.6%**
difference. The phase-pinned mean is a valid stand-in for a run that advances the
cursor, without paying for the cursor update.

## Measured

H200 NVL, the benchmark's own default config (`HV=32, H=16, K=V=128`, bf16 I/O,
fp32 state), kernels imported from `main` at 54cadad.

**Which phase actually flushes.** One step at each pinned `write_pos`, diffing
the checkpoint tensor, `L=8`:

```
write_pos=0..6   checkpoint written: False
write_pos=7      checkpoint written: True
```

A flush happens only at `write_pos == L-1`, as documented. The benchmark pins
`write_pos = 0`, the one phase that never flushes.

**What that costs.** Timing each phase separately, `B=256`:

```
L=8   0.1613 0.1672 0.1685 0.1691 0.1697 0.1705 0.1713 0.3790
                                                        ^^^^^^ the flush step
```

The flush step costs **2.3x** a non-flush step. The benchmark reports 0.1613 ms
and never shows this, so a cost that is bursty by construction is presented as
uniform.

**Across the batch sweep the benchmark itself runs:**

| B | L | packed | reported | true mean | speedup reported | speedup true |
|---:|---:|---:|---:|---:|---:|---:|
| 1 | 8 | 0.0088 | 0.0106 | 0.0112 | 0.83x | 0.78x |
| 1 | 16 | 0.0088 | 0.0102 | 0.0110 | 0.86x | 0.80x |
| 16 | 8 | 0.0279 | 0.0199 | 0.0217 | 1.40x | 1.29x |
| 16 | 16 | 0.0279 | 0.0199 | 0.0211 | 1.40x | 1.32x |
| 64 | 8 | 0.0782 | 0.0535 | 0.0612 | 1.46x | 1.28x |
| 64 | 16 | 0.0782 | 0.0528 | 0.0588 | 1.48x | 1.33x |
| 256 | 8 | 0.2813 | 0.1612 | 0.1948 | **1.75x** | **1.44x** |
| 256 | 16 | 0.2813 | 0.1608 | 0.1843 | **1.75x** | **1.53x** |

Latencies in ms; "true mean" is the mean over all L phases.

### L=8 and L=16 are indistinguishable in the reported column

At every batch size the reported speedups for `L=8` and `L=16` are the same to
within noise (1.40/1.40, 1.46/1.48, 1.75/1.75). Corrected, they separate:
1.29/1.32, 1.28/1.33, **1.44/1.53**.

The cause is that `BC = max(16, next_power_of_2(max_cache_len))` (line 577) is 16
for every `L <= 16`, so the non-flush step costs the same regardless of L, and
only including the flush phase captures that a larger L amortizes it over more
steps. Timing the non-flush step directly, `B=256`:

| L | BC | non-flush step |
|---:|---:|---:|
| 2 | 16 | 0.1601 ms |
| 4 | 16 | 0.1611 ms |
| 8 | 16 | 0.1614 ms |
| 16 | 16 | 0.1609 ms |
| 32 | 32 | 0.1777 ms |

Flat to 0.9% for every `L <= 16`, then a 10% jump when BC doubles. Whether that
separation should change anyone's choice of L is a system-level question this
benchmark cannot answer, since a larger L also doubles the ring's memory
footprint. The point here is only that the reported column hides a difference
that exists.

### Separately: `nk` is special-cased at L=1

Line 99 passes `nk = 1 if L == 1 else 2`. In the wrapper, `nk` is the number of
K-chunks the reconstruction walks: `BKT = BK // nk` with `BK =
next_power_of_2(K)`, validated by `BK % nk == 0` and `BKT >= 16` (lines 568-572).
At `K=128` both `nk=1` (`BKT=128`) and `nk=2` (`BKT=64`) pass that validation, so
`nk=2` is legal at `L=1`; there is no structural reason for the special case.

It is also the slower of the two. At `B=256`, `L=1`:

| | latency | vs packed |
|---|---:|---:|
| packed decode | 0.2812 ms | |
| ReplaySSM L=1, `nk=1` (what the benchmark runs) | 0.7336 ms | 0.38x |
| ReplaySSM L=1, `nk=2` | 0.3674 ms | 0.77x |

The special case doubles the reported L=1 cost, which is what produces the 0.38x
column. Using the same `nk` for all L would make that row comparable to the
others.

## Reproduction

```bash
python benchmark/kernels/attention/bench_gdn_replayssm_decode.py
```

Then, to see that no flush occurs, run one step per pinned `write_pos` and diff
the checkpoint:

```python
import torch
from sglang.kernels.ops.attention.fla.fused_recurrent_linear_replayssm import (
    fused_recurrent_gdn_replayssm_decode,
)

B, H, HV, K, V, L = 256, 16, 32, 128, 128, 8
dev, dt = "cuda", torch.bfloat16
inp = dict(
    mixed_qkv=torch.randn(B, 2 * H * K + HV * V, device=dev, dtype=dt),
    a=torch.randn(B, HV, device=dev, dtype=dt) * 0.5,
    b=torch.randn(B, HV, device=dev, dtype=dt),
    A_log=(torch.randn(HV, device=dev, dtype=torch.float32) * 0.3).contiguous(),
    dt_bias=(torch.randn(HV, device=dev, dtype=torch.float32) * 0.1).contiguous(),
    scale=K**-0.5,
    ssm_state_indices=torch.arange(B, device=dev, dtype=torch.int32),
    use_qk_l2norm_in_kernel=True,
    d_cache=torch.zeros(B, HV, L, V, device=dev, dtype=dt),
    k_cache=torch.zeros(B, H, L, K, device=dev, dtype=dt),
    g_cache=torch.zeros(B, HV, L, device=dev, dtype=torch.float32),
)
out = torch.empty(B, 1, HV, V, device=dev, dtype=dt)
base = torch.randn(B, HV, V, K, device=dev, dtype=torch.float32)

for p in range(L):
    state = base.clone()
    wp = torch.full((B,), p, device=dev, dtype=torch.int32)
    fused_recurrent_gdn_replayssm_decode(
        initial_state=state, out=out, write_pos=wp, nk=2, **inp
    )
    torch.cuda.synchronize()
    print(f"write_pos={p}  checkpoint written: {not torch.equal(state, base)}")
```

```
write_pos=0  checkpoint written: False
...
write_pos=7  checkpoint written: True
```

## Suggested fix

Time each of the L phases separately and report the mean, which measures the real
amortized cost without adding a cursor-update launch to the timed region. I have
this working and will open a PR referencing this issue.

## Scope of verification

- Single H200 NVL, `HV=32, H=16, K=V=128`, bf16 I/O with an fp32 state, i.e. the
  benchmark's own defaults. Not swept over head counts or state dtype.
- Kernels imported from the source tree at `main` 54cadad.
  `fused_recurrent_linear_replayssm.py` is unchanged from the released 0.5.18
  wheel; the packed-decode kernel's grid was reshaped since then (`(NV, B*HV)` to
  `(NV, B, HV)`) with no measurable effect at these shapes (packed 0.2805 ms vs
  0.2809 ms on the two versions).

## Environment

```
Python: 3.12.14 | packaged by conda-forge | (main, Aug 21 2026, 22:48:28) [GCC 14.4.0]
CUDA available: True
GPU 0: NVIDIA H200 NVL
GPU 0 Compute Capability: 9.0
CUDA_HOME: /usr/local/cuda-13.0
NVCC: Cuda compilation tools, release 13.0, V13.0.88
CUDA Driver Version: 580.173.02
PyTorch: 2.13.0+cu130
sglang: 0.5.18   (kernels imported from the source tree at main 54cadad)
sglang-kernel: 0.4.6.post1
flashinfer_python: 0.6.17
triton: 3.7.1
transformers: 5.12.1
numpy: 2.3.5
```

## One observation, not investigated

While validating the phase timings: the flush step costs **1.31x** packed decode
at `L=1`, where the ring is empty and the kernel header (lines 45-47) states the
kernel "reduces *algebraically* to the corresponding packed-decode kernel".
Holding `nk=2` and timing only the flush phase:

| L | real ring entries | flush step | vs packed |
|---:|---:|---:|---:|
| 1 | 0 | 0.3672 ms | 1.31x |
| 2 | 1 | 0.3681 ms | 1.31x |
| 8 | 7 | 0.3763 ms | 1.34x |
| 16 | 15 | 0.3874 ms | 1.38x |
| 32 | 31 | 0.4271 ms | 1.52x |

The cost does scale with the number of real ring entries, so the 1.31x at zero
entries is a fixed floor rather than fold work, and it is not explained by the BC
padding. I did not chase the cause; it is out of scope for the benchmark fix and
may well be expected.

### PR DESCRIPTION
**Pull request by davidli1515**
# [Bench] Amortize the GDN ReplaySSM decode latency over the flush cycle

Fixes #38398

## Motivation

`bench_gdn_replayssm_decode.py` creates `write_pos = torch.zeros(B)` and never
advances it, though the kernel wrapper's docstring says the caller must. The
kernel folds the ring into the checkpoint only when `write_pos == L - 1`, so with
the cursor pinned at 0 and `L >= 2` the timed step never flushes and never
replays: every iteration re-measures the cheapest phase of the L-phase cycle.

The two reported columns therefore disagree. `state-traffic` is computed from a
model that assumes one checkpoint write per L steps:

```python
replay = (state_elems * fp32) * (1 + 1.0 / L)   # read every step + write 1/L
```

while `speedup` is measured on a path that performs that write zero times per L
steps. Both labels claim the amortized quantity.

## Modifications

**Time both phases and combine them.** A steady-state cycle is `L-1` non-flush
steps and one flush step, so the benchmark now times one of each and reports

```
((L - 1) * t_nonflush + t_flush) / L
```

The non-flush steps are not identical (each replays one more ring entry than the
last), so the *midpoint* phase is sampled as their representative. Phase 0
replays nothing and is the cheapest step in the cycle, which is exactly what made
the old number wrong.

Sweeping every phase and comparing (see the table below) puts this estimator
within **0.5%** of the exact mean, at two timings per L instead of L. A full
sweep would make the benchmark 8-16x slower to run for a sub-percent gain.

**Report the split.** The flush step costs ~2.3x a non-flush one, so the per-step
mean alone still hides that the cost is bursty by construction. Each row now
prints both components.

**Drop the `nk = 1 if L == 1 else 2` special case**, and expose `--nk` instead.
`nk` is the number of K-chunks the reconstruction walks (`BKT = BK // nk`,
validated by `BK % nk == 0` and `BKT >= 16`); at `K=128` both values are legal at
every L, so the special case was not structural, and `nk=1` is the slower of the
two by 2x. It was what produced the old 0.38x column at L=1.

## Results

H200 NVL, the benchmark's own defaults (`HV=32, H=16, K=V=128`, bf16 I/O, fp32
state).

| B | L | packed | before | after | speedup before | speedup after |
|---:|---:|---:|---:|---:|---:|---:|
| 1 | 8 | 0.0088 | 0.0106 | 0.0113 | 0.83x | 0.78x |
| 1 | 16 | 0.0088 | 0.0102 | 0.0110 | 0.86x | 0.80x |
| 16 | 8 | 0.0277 | 0.0199 | 0.0218 | 1.40x | 1.27x |
| 16 | 16 | 0.0277 | 0.0199 | 0.0213 | 1.40x | 1.30x |
| 64 | 8 | 0.0775 | 0.0535 | 0.0610 | 1.46x | 1.27x |
| 64 | 16 | 0.0775 | 0.0528 | 0.0589 | 1.48x | 1.32x |
| 256 | 8 | 0.2808 | 0.1612 | 0.1955 | **1.75x** | **1.44x** |
| 256 | 16 | 0.2808 | 0.1608 | 0.1847 | **1.75x** | **1.52x** |

Latencies in ms. The old column overstated the speedup by 6-21% depending on
batch size.

`L=8` and `L=16` were indistinguishable before (1.40/1.40, 1.46/1.48, 1.75/1.75)
because the non-flush step costs the same for both: `BC = max(16,
next_power_of_2(max_cache_len))` is 16 for every `L <= 16`. Only including the
flush phase shows that a larger L amortizes it over more steps.

## Validation of the estimator

Every phase timed for each (B, L) the benchmark reports, exact mean against the
shipped two-point estimate and against the two obvious wrong choices:

| B | L | exact mean | 2pt (mid) | err | 2pt (phase 0) | err | old 1pt | err |
|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| 1 | 8 | 0.0112 | 0.0111 | -0.5% | 0.0111 | -0.8% | 0.0106 | -5.6% |
| 1 | 16 | 0.0110 | 0.0110 | -0.0% | 0.0105 | -4.8% | 0.0102 | -7.5% |
| 16 | 8 | 0.0218 | 0.0218 | +0.0% | 0.0216 | -1.0% | 0.0201 | -8.0% |
| 16 | 16 | 0.0212 | 0.0213 | +0.2% | 0.0207 | -2.4% | 0.0200 | -6.1% |
| 64 | 8 | 0.0611 | 0.0612 | +0.2% | 0.0602 | -1.4% | 0.0534 | -12.6% |
| 64 | 16 | 0.0588 | 0.0591 | +0.5% | 0.0564 | -4.0% | 0.0528 | -10.1% |
| 256 | 8 | 0.1939 | 0.1947 | +0.4% | 0.1879 | -3.1% | 0.1611 | -17.0% |
| 256 | 16 | 0.1838 | 0.1843 | +0.3% | 0.1746 | -5.0% | 0.1606 | -12.6% |

Worst error of the shipped estimator: **0.5%**. Using phase 0 as the non-flush
representative instead would cost up to 5.0%, which is why the midpoint is
sampled.

Independently, a loop that really advances `write_pos` every step costs **0.1935
ms/step** at `B=256, L=8`, against a phase-pinned mean of **0.1946 ms/step**, a
0.6% difference. Pinning the cursor and averaging measures the same thing as
cycling it, without folding a per-step cursor-update launch into the timing.

## Sample output

```
GDN ReplaySSM decode microbench  HV=32 H=16 K=128 V=128 dtype=bf16 nk=2
replay = mean per-step latency over one L-phase cycle (ms), split into its
non-flush and flush steps; speedup = packed/replay; state-traffic = replay/packed
(lower is better)

B=256   packed=0.2806 ms
    L=1   replay=0.3667 ms  speedup= 0.77x  state-traffic= 1.00x  (flush only)
    L=8   replay=0.1955 ms  speedup= 1.44x  state-traffic= 0.56x  (non-flush 0.1693, flush 0.3791)
    L=16  replay=0.1848 ms  speedup= 1.52x  state-traffic= 0.53x  (non-flush 0.1713, flush 0.3873)
```

## Checklist

- [x] Format the code and run the linters
- [x] Add unit tests / benchmarks where applicable — this is a benchmark-only
      change; the estimator it introduces is validated against a full per-phase
      sweep in the table above
- [x] Update documentation — the module docstring now describes what is timed and
      why

## Scope

- Benchmark script only. No runtime code is touched, so there is no correctness
  or performance impact on serving.
- Measured on a single H200 NVL at the benchmark's own defaults, kernels from
  `main`. Not swept over head counts or state dtype.
- Runtime cost of the change: two timings per L instead of one, so the benchmark
  takes about twice as long to run.

## One observation, not addressed here

The flush step costs **1.31x** packed decode at `L=1`, where the ring is empty
and the kernel header states it "reduces *algebraically* to the corresponding
packed-decode kernel". Holding `nk=2` and timing only the flush phase:

| L | real ring entries | flush step | vs packed |
|---:|---:|---:|---:|
| 1 | 0 | 0.3672 ms | 1.31x |
| 2 | 1 | 0.3681 ms | 1.31x |
| 8 | 7 | 0.3763 ms | 1.34x |
| 16 | 15 | 0.3874 ms | 1.38x |
| 32 | 31 | 0.4271 ms | 1.52x |

The cost does scale with the number of real ring entries, so the 1.31x at zero
entries is a fixed floor rather than fold work, and it is not explained by the BC
padding. I did not chase the cause; it is out of scope for a benchmark fix and
may well be expected.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #34182194168](https://github.com/sgl-project/sglang/actions/runs/34182194168)
Latest PR Test (Extra): :no_entry_sign: [Run #34200835081](https://github.com/sgl-project/sglang/actions/runs/34200835081)
Latest PR Test (AMD ROCm 7.2): :heavy_minus_sign: **No AMD PR run found for this commit**.

### CODE REVIEW

**Review by yuan-luo — APPROVED** [review]
LGTM

**Comment by yuan-luo** [comment]
All GPU CI passed.
